### ISSUE TITLE
Protecting user anonymity

### ISSUE THREAD

**Issue by john-light** [argentlabs/hopper#1]
Hello, please excuse me if this is the wrong place to ask. How is the user's anonymity protected from the Argent server (mixer / ethereum node) with this software? It seems that as implemented the server can trivially de-anonymize all users.

**Comment by HarryR** [comment]
Given that the zkSNARK proofs are generated on your own device there is no secret leakage.

Relayers exists to prevent on-chain correlation of deposit & withdraw actions from the individual accounts which are making them.

However, if you relay the transaction through Tor, for example, to a relayer which takes a small cut of the transaction (enough to cover Gas costs), then they don't know your IP and neither them nor anybody observing the on-chain data can associate a deposit with a withdraw.

The only other way of correlating is if you withdraw two coins to the same account, or whatever you do after withdrawing them which could de-anonymise you.

At the moment, it doesn't seem like transactions are submitted via Tor or a source-IP anonymization service, so relayers can observe the IP addresses - if they act honestly then the on-chain privacy is maintained.

ZCash and Monero have similar problems, if you are able to observe the IP addresses of people submitting transactions, you can de-anonymise them.

**Comment by john-light** [comment]
> The only other way of correlating is if you withdraw two coins to the same account, or whatever you do after withdrawing them which could de-anonymise you.

Right, and this includes withdrawing to the same wallet - even if a different account - if the wallet uses the same third-party node to sync the balances of each account. It may be instructive to see [how Wasabi Wallet solves this issue](https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Documentation/FAQ.md#how-do-i-connect-my-own-full-node-to-wasabi) using Tor, client-side block filtering (BIP 157 + 158), and the option for a user to connect to their own node rather than using a third party node. 

If strong correlation protection is not in place for all users, by default, then this shrinks the anonymity set and makes it easier to correlate even users that take advantage of all correlation-protection measures with their transactions.

**Comment by kayagoban** [comment]
If the user is running his own full node, how much does this mitigate the problem of address  correlation?

**Comment by kayagoban** [comment]
> If the user is running his own full node, how much does this mitigate the problem of address correlation?

To be specific, I'm looking at including native support for hopper in Shadowlands (https://github.com/kayagoban/shadowlands)

**Comment by HarryR** [comment]
There are three side channels which you can use to obtain information about somebody making a proof:

 1) Ethereum address that the transaction is submitted from
 2) Information required to create the proof, retrieved from the blockchain (e.g. calculating your path in a merkle tree)
 3) IP address (or other identifiable client information) which may be observed, e.g. malicious IP-logging relayer or targeted interception.

Running a full node helps with 2). However, for 1) you need somebody else to submit the transaction for you, and 3) is a problem for either.

**Comment by john-light** [comment]
> If the user is running his own full node, how much does this mitigate the problem of address  correlation?

@kayagoban every user would need to use their own full node or a Wasabi-like light client connecting over Tor by default so they are all protected from de-anonymization and included in the anonymity set.

**Comment by kayagoban** [comment]
My particular case would be the Shadowlands program, which runs in a unix terminal and  requires access to a full node to run.

C Thomas
Soykaf Digital Systems LLC
ph. +14798568793
cthomas@soykaf.digital

**Comment by leafcutterant** [comment]
> if you relay the transaction through Tor,

@HarryR This is paramount. If no obfuscation is used on the transfer layer, *or* the same obfuscation path is used for depositing and withdrawing, the relayer (or an apparent group of them -- it's easy to sybil) has all the information to deanonymize participants.

Is Tor built into the application? Or at least is it easily routable through Tor?

I'd say baking it into the application (like Bisq does) would be a huge boost.
